### PR TITLE
Add buildCallWithPermit variations

### LINKED ISSUE #323: SDK V3: Permit2 - Add buildCallWithSignatures
Add one extra buildCall function that expects Permit2 signatures as input and returns the final encoded callData.

This applies to the following operations:
- Add Liquidity
- Remove Liquidity
- Swap
- BatchSwap
- Init (?)
- Create (?)

### PR DESCRIPTION
**Pull request by brunoguerios**
Close #323 

Main changes:
- swap with permit2 signatures
- add liquidity with permit2 signatures
- remove liquidity with permit signatures

Also:
- `approveTokens` and `setTokenBalances` were exposed from test helpers, so we have more flexibility when setting up a test
- add/remove/swap test helpers now have an extra `usePermitSignatures` flag
- add/remove/swap tests were split into `permit direct approval` and `permit signatures` which basically replicate tests for those 2 conditions, so we make sure we’re not missing any scenario that needs to be tested
- new input validator for buildCallWithPermit was added

Notes:
- addLiquidity and swap when `wethIsEth = true` is currently not supported by SC (may be fixed in the future, but I suggest we leave it out of scope for this PR)

### CODE REVIEW

File: src/entities/inputValidator/inputValidator.ts
Comment (brunoguerios) [inline src/entities/inputValidator/inputValidator.ts:99]: This is not following the pattern, but it simplifies the implementation significantly.
To be honest, I don't like the way InputValidator is currently implemented. Seems to be more complicated than it needs to be.
In any case, I think it's ok as it, but let me know what you think

Comment (MattPereira) [reply src/entities/inputValidator/inputValidator.ts:99]: Definitely agree `InputValidator` could be done better and I like the static method pattern you're proposing here. Much easier to read :+1:

File: src/entities/permit2/domain.ts
Comment (MattPereira) [inline src/entities/permit2/domain.ts:20]: Is this type declaration being used?

File: src/entities/inputValidator/inputValidator.ts
Comment (johngrantuk) [reply src/entities/inputValidator/inputValidator.ts:99]: I don't think I fully appreciate the complexity of following the old pattern here but open to the change if we think it improves things.
If we want to move to a static refactor lets make an issue for it.

File: src/entities/permit/index.ts
Comment (johngrantuk) [inline src/entities/permit/index.ts:66]: I think we can tidy this up a bit. Doesn't look like we ever expect nonce to be passed (e.g. in signRemoveLiquidityApproval). 
I think we should:
* move `getNonce` out of `signPermit` and into `signRemoveLiquidityApproval`
then either:
* change `signPermit` nonce to be non-optional or
* change `signRemoveLiquidityApproval` to accept an optional nonce

File: src/entities/permit/index.ts
Comment (johngrantuk) [inline src/entities/permit/index.ts:1]: How about renaming parent folder to `PermitHelper` which keeps with the general standard we use?
Also I think there's some stuff being exported that don't need to be?

File: src/entities/permit2/allowanceTransfer.ts
Comment (johngrantuk) [inline src/entities/permit2/allowanceTransfer.ts:67]: Never seen this type of thing before: `permit is PermitSingle` seems like its just another way of saying `boolean` or is there a specific reason for it?

File: src/entities/swap/swaps/v3/index.ts
Comment (johngrantuk) [inline src/entities/swap/swaps/v3/index.ts:613]: Do you think this is actually useful? I find it less explicit.

File: test/lib/utils/helper.ts
Comment (johngrantuk) [inline test/lib/utils/helper.ts:1]: While we are refactoring this how about also updating the `approveSpender` type functions to first read `hasApprovedToken` and only write the approval if needed?

**Review by johngrantuk — CHANGES_REQUESTED** [review]
So nice to see this all coming together and tests passing!
Few comments/changes requested, also:
* we'll need to add some info to the docs about the various permit nuances. Can you create an issue for this (that includes the relevant TLDR) and make sure its added to SDK project
* In the main comment it says:
"addLiquidity and swap when wethIsEth = true is currently not supported by SC (may be fixed in the future, but I suggest we leave it out of scope for this PR)"
- but it looks like swap tests do have wethIsEth=true. Can you explain a bit more detail about what is/isn't supported.

File: src/entities/inputValidator/inputValidator.ts
Comment (brunoguerios) [reply src/entities/inputValidator/inputValidator.ts:99]: This buildCallWithPermit input validation is also used by swaps, which don't follow the InputValidator structure. 
The static alternative was an attempt to get around the overhead of that structure and still keep things close together.
There is already an issue about refactoring input validators [here](https://github.com/balancer/b-sdk/issues/280). We could/should expand on that and try to find a new design that is simpler/easier to follow and also a bit more flexible.
It doesn't even need to follow that static pattern if we find another solution that works better considering the whole implementation.
Let me know if you think that's not enough reason to break pattern and I'll adjust over here until we spend proper time on this later on.

File: src/entities/permit2/domain.ts
Comment (brunoguerios) [reply src/entities/permit2/domain.ts:20]: Apparently not 😅 I adapted the implementation from another repo and I forgot to delete it.
Will remove 👍

File: src/entities/permit/index.ts
Comment (brunoguerios) [reply src/entities/permit/index.ts:66]: Yeap, that makes sense 👍

File: src/entities/swap/swaps/v3/index.ts
Comment (brunoguerios) [reply src/entities/swap/swaps/v3/index.ts:613]: Yeah, I changed because I found myself repeating the code of checking for `isBatchSwap` and assigning `to` to the proper address, but I agree that having these things closer to where it's being used is more explicit.
I'll rollback that change 👍

File: src/entities/permit/index.ts
Comment (brunoguerios) [reply src/entities/permit/index.ts]: A natural extension of this logic is to allow for `nonce` and `deadline` to be provided as input on permit2. The implementation is a bit more complex because user has to provide these configs for each tokenIn. I tried to keep it as flexible as possible. Let me know what you think.

**Comment by brunoguerios** [comment]
I agree that it would be nice to document Permit/Permit2 signatures nuances. I created #369 so we don't forget to do it. Just to be clear, the idea is to add content to readme (including eventual limitations, such as `does not support wethIsEth = true`) and also some examples on how to use it, right?
Ps: I'll wait on a final decision from the SC team to add details about what's supported and what's not - if I understood correctly, everything will be supported in the next testnet deployment.

File: src/entities/inputValidator/inputValidator.ts
Comment (johngrantuk) [reply src/entities/inputValidator/inputValidator.ts:99]: All good for now I think.

**Comment by johngrantuk** [comment]
> I agree that it would be nice to document Permit/Permit2 signatures nuances. I created #369 so we don't forget to do it. Just to be clear, the idea is to add content to readme (including eventual limitations, such as `does not support wethIsEth = true`) and also some examples on how to use it, right? Ps: I'll wait on a final decision from the SC team to add details about what's supported and what's not - if I understood correctly, everything will be supported in the next testnet deployment.

Replying on the issue directly.

**Review by johngrantuk — APPROVED** [review]
Lets release this beast!!

**Pull request merged**
